src/utils/helpers.py

```diff
@@ -2,9 +2,67 @@
 import logging
 from pathlib import Path
 
+import yaml
+from pyspark.sql import DataFrame
+from pyspark.sql.types import DecimalType, StringType, TimestampType
+
 PATH_SPARK_LOG = "/tmp/logs/spark.log"
 
 
+def convert_datetime_columns(
+    df: DataFrame, ignore_cols: list[str]
+) -> DataFrame:
+    """
+    iasWorld stores datetimes via the TIMESTAMP data type, but historically
+    we converted these columns to strings to avoid messy type incompatibility
+    with Athena. This function converts TIMESTAMP columns to UTC datetime
+    strings. Columns specified via schema overrides will ignore this conversion.
+
+    Args:
+        df: Spark DataFrame with columns to convert.
+        ignore_cols: Ignore columns specified in this list during conversion.
+
+    Returns:
+        Spark DataFrame with TIMESTAMP columns converted to STRING.
+    """
+    for field in df.schema.fields:
+        if (
+            isinstance(field.dataType, TimestampType)
+            and field.name not in ignore_cols
+        ):
+            df = df.withColumn(field.name, df[field.name].cast(StringType()))
+    return df
```

Comment:
**[Thought, non-blocking]** This might just be a personal preference, but I dislike the pandas convention of defining single-use functions that take in a dataframe, transform a bunch of the dataframe's columns, and then return a new copy of the dataframe with transformed columns. I think the intention behind the pattern is to make the top-level script easier to read by summarizing each transformation and factoring it out into a separate function, but as a reader I tend to find the pattern _more_ confusing than the alternative because the lack of references to column names or types means that the function interface doesn't actually tell you what columns are undergoing transformation or what that transformation should look like.

Still, I recognize that this is a common pandas pattern, and I'm a lot newer to writing production pandas than you are! I'm just curious if you've heard this critique before and if you have thoughts about the right approach to readability for complex pandas transformations. My impression is that R solves this problem with the tidyverse, which provides a bunch of nice abstractions around data transformations that make it unnecessary to factor logic out into isolated functions in the majority of cases. I'm not sure if there's an equivalent approach to pandas code but I'd be interested in learning more about it.

Reply:
I also dislike it and would usually in-line code like this, but in this case I wanted to call out and document what's happening since it's pretty important/weird behavior. I think instead I'll move these inline and add a note to the README then link to it.